README.md

```diff
@@ -1,77 +1,150 @@
-regards to parameters that may need to be set for other roles, or variables
-that are used from other roles.
+Set this variable to `true` to indicate that you accept EULA for using the
+`mssql-conf` utility.
 
-## Example Playbook
+The license terms for this product can be found in `/usr/share/doc/mssql-server`
+or downloaded from <https://go.microsoft.com/fwlink/?LinkId=2104078&clcid=0x409>.
+The privacy statement can be viewed at
+<https://go.microsoft.com/fwlink/?LinkId=853010&clcid=0x409>.
+
+Default: `false`
+
+### `mssql_password`
+
+The password for the database sa user. The password must have a minimum length
+of 8 characters, include uppercase and lowercase letters, base 10 digits or
+non-alphanumeric symbols. Do not use single quotes ('), double quotes ("), and
+spaces in the password because `sqlcmd` cannot authorise when the password
+includes those symbols.
+
+When running this role on a host that has MSSQL set up, the mssql_password
+variable overwrites the existing sa user password to the one that you specified.
+
+Default: `None`
+
+### `mssql_edition`
+
+The edition of MSSQL to install. Use one of the following values:
+
+* `Enterprise`
+* `Standard`
+* `Web`
+* `Developer`
+* `Express`
+* `Evaluation`
+* A product key in the form `#####-#####-#####-#####-#####`, where `#` is a
+  number or a letter.
+  For more information, see
+  <https://docs.microsoft.com/en-us/sql/linux/sql-server-linux-configure-environment-variables?view=sql-server-ver15>.
+
+Default: `None`
+
+### `mssql_tcp_port`
+
+The port that MSSQL listens on.
+
+Default: `1443`
+
+### `mssql_ip_address`
 
-Including an example of how to use your role (for instance, with variables
-passed in as parameters) is always nice for users too:
+The IP address that MSSQL listens on.
+
+WARNING: The role applies the same IP address setting to all hosts that you run
+the role against. You cannot specify a different IP address for a different
+host. Therefore, you must set this variable only when running the role against
+one host.
+
+Default: `0.0.0.0`
+
+### `mssql_input_sql_file`
+
+You can use the role to input a file containing SQL statements or procedures into
+MSSQL. With this variable, enter the path to the SQL file containing the
+database configuration.
+
+When specifying this variable, you must also specify the `mssql_password`
+variable because authentication is required to input an SQL file to MSSQL.
+
+If you do not pass this variable, the role only configures the MSSQL Server
+and does not input any SQL file.
+
+Note that this task is not idempotent, the role always inputs an SQL file if
+this variable is defined.
+
+This is an example of the SQL file content:
+
+```sql
+CREATE DATABASE ExampleDB
+GO
+USE ExampleDB
+CREATE TABLE Inventory (id INT, name NVARCHAR(50), quantity INT)
+INSERT INTO Inventory VALUES (1, 'apple', 100)
+INSERT INTO Inventory VALUES (2, 'orange', 150)
+INSERT INTO Inventory VALUES (3, 'banana', 154)
```

Comment:
Multi-byte names are acceptable?

Reply:
We have the same problem with every other role - that is, how do we know that when the user inputs some Unicode, the underlying program being managed will accept it?  So I think we should not worry about this for now.

Reply:
I agree. I just think our customers would put words in their language when they create the initial db. Compared to the other roles, database is more closely related to their own business. So, I think the chance to see the non-ascii chars would be higher. And I also think there shouldn't be any problem doing that in the mssql env in rhel/fedora. So, as long as we (ansible / system-roles) do not do anything extra, it's supposed to work. :) My hope is when we develop test cases, we could add such characters to the test script and check they are not corrupted (if possible).

Reply:
Might be an ansible output issue, but if I changed "banana" with "バナナ" in tests/sql_script.sql, it's not shown properly in the stdout. My LANG env var is `LANG=en_US.UTF-8`.
```
TASK [linux-system-roles.mssql : Input the sql_script.sql file with the sqlcmd command] *****************
task path: /home/nhosoi/linux-system-roles/mssql/tasks/main.yml:279
changed: [/home/nhosoi/Downloads/CentOS-8-GenericCloud-8.3.2011-20201204.2.x86_64.qcow2] => {
    "changed": true,
    "cmd": [
        "/opt/mssql-tools/bin/sqlcmd",
        "-U",
        "sa",
        "-P",
        "p@55w0rD",
        "-i",
        "/tmp/ansible.hn855rap"
    ],
    "delta": "0:00:00.849502",
    "end": "2021-03-19 02:50:51.448416",
    "rc": 0,
    "start": "2021-03-19 02:50:50.598914"
}
STDOUT:
Changed database context to 'ExampleDB'.
(1 rows affected)
(1 rows affected)
(1 rows affected)
id          name                                               quantity   
----------- -------------------------------------------------- -----------
          2 orange                                                     150
          3 ???                                                        154
(2 rows affected)
```

Reply:
I think you are raising a very good point here @nhosoi. 
I am not an expert with SQL but I think that some additional configuration is needed to use a different (not the default English) language with a database.
I think changing the database language is a separate task and is referred to as `changing the locale` in MSSQL context: https://docs.microsoft.com/en-us/sql/linux/sql-server-linux-configure-mssql-conf?view=sql-server-ver15#lcid
Now, this role does not do such functionality, so I think it defaults to using English.
Let me test using a different language with MSSQL. I think that this is an important function and would make the role more inclusive.

Reply:
I hope that Ansible will be able to process "バナナ" normally too.

Reply:
So, Amit clarified that MSSQL supports Unicode characters by default, you just need to use a different syntax - add `N` prior to the string containing them. Like that:
`INSERT INTO Inventory VALUES (4, N'バナナ', 1)`
I have asked Amit to verify that changing locale is not something the basic role should do as seems like it works as is.

Reply:
@spetrosi, thank you for the clarification!! (on the other items, as well)

Yes, once I added N to the multi-byte word, it worked just fine including the output from ansible. It's great to know. :)

I had one more question in this thread.
> But the test script tests_input_sql_file.yml says it was successful... Could there be any way to detect such a failure?

Is it possible to check the result of the task "Input the sql_script.sql file with the sqlcmd command" and issue an error if it fails?

Reply:
I think this is a good idea. The role will register the output of the sqlcmd input and fail if this task had an error in it. I will see how to implement this better next week.

Reply:
hmm - so https://github.com/linux-system-roles/mssql/blob/0105ebf61ca400ab272109a124bdc25deccdef1c/tasks/main.yml#L279 - the `sqlcmd` does not exit with a non-zero exit code in this case?  Maybe it thinks this is a "warning" case and does not exit with an error code? If so, is there some command line flag we can use to tell `sqlcmd` to "treat warning as error"?

Reply:
Unfortunately, sqlcmd always exits with `0`. Instead, it returns errors, messages, and states. Users are expected to proceed with those error messages themselves. There is a ton of different messages and again I do not think that it is effective to proceed with them.

For example, if items that the script adds already exists, the error message is the following:
```
Msg 2714, Level 16, State 6, Server ibm-p8-kvm-03-guest-02, Line 2
There is already an object named 'Inventory' in the database.
```

If I introduce an error in the syntax of "GO":
```
Msg 102, Level 15, State 1, Server ibm-p8-kvm-03-guest-02, Line 7",
"Incorrect syntax near 'GrO'."
```

If I introduce an error in the syntax of "USE":
```
Msg 2812, Level 16, State 62, Server ibm-p8-kvm-03-guest-02, Line 1
Could not find stored procedure 'USrE'.
SqlState 24000, Invalid cursor state
```

Note that `SqlState 24000` is returned when the syntax of `USE` is incorrect, but is not returned when the syntax of `GO` is incorrect.

When everything is fine and sqlcmd succeeds to input an SQL file without any issues, it won't print any messages at all. So if the string "Msg" is present in the output, it means that something went wrong. At the same time, Msg must not fail the playbook, as it might be that a unit (table or value) already exists within the database. We could have marked the task as Warning when the string "Msg" is in it, but this will require creating a module.
Just thinking out loud, if we knew the IDs of Msg that are actual errors (should result in an error), and warnings (should result in a Warning), we could process this to make the module more efficient. But if such a list does not exist already, it would be a big PITA to collect it.

Reply:
Unfortunately, sqlcmd always exits with `0`. Instead, it returns errors, messages, and states. Users are expected to proceed with those error messages themselves. There is a ton of different messages and again I do not think that it is effective to proceed with them.

For example, if items that the script adds already exists, the error message is the following:
```
Msg 2714, Level 16, State 6, Server ibm-p8-kvm-03-guest-02, Line 2
There is already an object named 'Inventory' in the database.
```

If I introduce an error in the syntax of "GO":
```
Msg 102, Level 15, State 1, Server ibm-p8-kvm-03-guest-02, Line 7",
"Incorrect syntax near 'GrO'."
```

If I introduce an error in the syntax of "USE":
```
Msg 2812, Level 16, State 62, Server ibm-p8-kvm-03-guest-02, Line 1
Could not find stored procedure 'USrE'.
SqlState 24000, Invalid cursor state
```

Note that `SqlState 24000` is returned when the syntax of `USE` is incorrect, but is not returned when the syntax of `GO` is incorrect.

When everything is fine and sqlcmd succeeds to input an SQL file without any issues, it won't print any messages at all. So if the string "Msg" is present in the output, it means that something went wrong. At the same time, Msg must not fail the playbook, as it might be that a unit (table or value) already exists within the database. We could have marked the task as Warning when the string "Msg" is in it, but this will require creating a module.

Reply:
This sounds like a question for the microsoft team - how can we tell, from an automated program, if `sqlcmd` failed or not?